game/src/plugins/skills/mining/src/mining.plugin.kts

```diff
@@ -0,0 +1,130 @@
+import org.apollo.cache.def.ItemDefinition
+import org.apollo.game.action.DistancedAction
+import org.apollo.game.message.impl.ObjectActionMessage
+import org.apollo.game.model.Position
+import org.apollo.game.model.entity.EquipmentConstants
+import org.apollo.game.model.entity.Player
+import org.apollo.game.model.entity.Skill
+import org.apollo.game.plugin.skills.mining.*
+import kotlin.properties.Delegates
+
+class MiningAction : DistancedAction<Player> {
+
+    var ore: Ore by Delegates.notNull()
+    var counter: Int = 0
+    var started: Boolean = false
+
+    constructor(player: Player, position: Position, ore: Ore) : super(0, true, player, position, 1 /* ORE SIZE */) {
+        this.ore = ore
+    }
+
+    override fun executeAction() {
+        val level = mob.skillSet.getSkill(Skill.MINING).currentLevel
+        val pickaxe = findPickaxe()
+        mob.turnTo(position)
+
+        //check that our pick can mine the ore
+        if (pickaxe == null || level < pickaxe.level) {
+            mob.sendMessage("You do not have a pickaxe for which you have the level to use.")
+            stop()
+            return
+        }
+
+        //check that we can mine the ore
+        if (ore.level > level) {
+            mob.sendMessage("You do not have the required level to mine this rock.")
+            stop()
+            return
+        }
+
+        //start the process of mining
+        if (started) {
+            if (counter == 0) {
+                //TODO: calculate the chance that the player can actually get the ore
+                if (mob.inventory.add(ore.id)) {
+                    mob.sendMessage("You managed to mine some " + ItemDefinition.lookup(ore.id).name.substring(3).toLowerCase() + ".")
+                    mob.skillSet.addExperience(Skill.MINING, ore.exp)
+                    //TODO: Expire ore
+                }
+            }
+            counter -= 1
+        } else {
+            startMine(pickaxe)
+        }
+    }
+
+    fun findPickaxe(): Pickaxe? {
+        var holding = mob.equipment.get(EquipmentConstants.WEAPON)
```

Comment:
I think this method needs to be a little smarter.  When you mine an ore I'm sure that you use the best pickaxe available to you.

Reply:
This function actually does that already. The picks that it iterates over are in order of best to worst.

Reply:
Yep, but it doesn't seem to take the players level into account.  i.e., if they have a rune pickaxe in their inventory but can only use adamant then findPickaxe() should return the adamant pickaxe.

Reply:
Oh, I forgot about that. Thank you!!